robotframework-ls/src/robotframework_ls/impl/semantic_tokens.py

```diff
@@ -97,9 +108,101 @@ def extract_library_token_from_keyword(keyword_token):
     library_token_length = keyword_token.value.rfind(".")
     if library_token_length > 0:
         library_token = keyword_token.extract_token("name", library_token_length)
+        library_token.index = TOKEN_TYPE_TO_INDEX[library_token.type]
     return library_token
 
 
+def extract_part_of_bdd_statement(keyword_token):
+    import re
+
+    bdd_statement_token = None
+    result = re.match(r"([^$&@]+)", keyword_token.value)
+    if result:
+        bdd_statement_value = result.group(1)
+        if bdd_statement_value[-1].isspace():
+            trailing_whitespace = 1
+        else:
+            trailing_whitespace = 0
+        bdd_statement_length = len(bdd_statement_value) - trailing_whitespace
+        bdd_statement_token = keyword_token.extract_token(
+            keyword_token.type, bdd_statement_length, trailing_whitespace
+        )
+        bdd_statement_token.index = RF_TOKEN_TYPE_TO_TOKEN_TYPE_INDEX[KEYWORD]
+    return bdd_statement_token
+
+
+def extract_embedded_variable(keyword_token):
+    import re
+
+    embedded_variable_token = None
+    result = re.match(r"([$&@]{[^}]*?})", keyword_token.value)
```

Comment:
This isn't entirely correct because variables may be nested.

i.e.: in the code below `${OUTPUT${DIR}}` and `${OUTPUT_DIR}` are the same thing...

```
*** Test Cases ***
My Test
    ${output}=    Set Variable    OUTPUT_
    ${dir}=    Set Variable    DIR
    Log To Console    ${OUTPUT${DIR}}
    Log To Console    ${OUTPUT_DIR}
```

You need to do something as:

```
    token = ast_utils.create_token(text)
    try:
        tokenized_vars = ast_utils.tokenize_variables(token)
        for v in tokenized_vars:
            if v.type == v.VARIABLE:
            ...
    except VariableError:  # May throw error if it's not valid
       ...
```

We can initially not do it recursively to get the sub-variables if you feel it's too much work to do the recursion and get sub-variables at this point, but we should try to structure keeping that in mind...